### DISCUSSION TITLE
Taking in large prompts (5000 characters and up)

### DISCUSSION THREAD

**Discussion by ActuallyTaylor** [ggml-org/llama.cpp#3704] [Q&A]
I am working on a project that requires running text generation on prompts of over 5000 characters. Currently loading them into a single tokenization call results in the fail state created in #1881. I believe this is because I am attempting to load significantly more tokens than what is supposed to be loaded at once.

I have looked over the examples provided but I have not found any examples of loading really large prompts. I have done a little bit of research into loading large prompts but I am admittedly very new to this field and was wondering if I could receive some guidance specific to this project. 

Through my research I have found techniques like chunking my input into smaller fragments, however I am not sure how I would implement this using the llama.cpp api. I already have actual string fragmentation completed but my question lies in how this would be sent to the model. Is this the right step forward or are there other resources / techniques that I should explore?

**Comment by shibe2** [comment]
I regularly use large prompts like 10000 characters. You should describe issues that you are encountering in more detail.

**Reply by ActuallyTaylor** [reply]
I can definitely go into more detail!

To start, my project is written in Swift, using the C++ and C interoperability released recently. I do not believe this will be the source of issues but I am willing to rewrite my code in C++ or C if anyone believes that may be an issue.

## My Implementation:
I am first taking in user input and then passing it into a tokenization function which uses `llama_tokenize` to tokenize the text. This appears to be working good as I receive **7128** tokens back after providing my input.

My next step is to construct a context using `llama_new_context_with_model` with a batch size and context size of **7128** + a given length of tokens to generate. I then create a batch using `llama_batch_init` I then set the proper references using the code found in [batched.swift](https://github.com/ggerganov/llama.cpp/blob/465219b9143ac01db0990bbcb0a081ef72ec2008/examples/batched.swift/Sources/main.swift#L80).

However here is where the crash / errors begin. Once I call `llama_decode` on my newly created batch and context I fall into what seems like a very long loop eventually landing at 

```
2023-10-20 21:55:27.126807-0400 xctest[4655:125069] Execution of the command buffer was aborted due to an error during execution. Internal Error (0000000e:Internal Error)
ggml_metal_graph_compute: command buffer 0 failed with status 5
GGML_ASSERT: MY_PATH/llama.cpp/ggml-metal.m:1459: false
```

I have tracked this error too the following lines of code in `ggml-metal.m`
```
// check status of command buffers
    // needed to detect if the device ran out-of-memory for example (#1881)
    for (int i = 0; i < n_cb; i++) {
        [ctx->command_buffers[i] waitUntilCompleted];

        MTLCommandBufferStatus status = (MTLCommandBufferStatus) [ctx->command_buffers[i] status];
        if (status != MTLCommandBufferStatusCompleted) {
            GGML_METAL_LOG_INFO("%s: command buffer %d failed with status %lu\n", __func__, i, status);
            GGML_ASSERT(false);
        }
    }
```

## End Notes
This specific error is weird to me as I am running on an Apple M1 Max with 64GB of memory, and 49152.00 MB for my `recommendedMaxWorkingSetSize`.

I am assuming there is a fundamental issue in my steps here which is causing this crash so please let me know! Also if there is any other context I can provide please let me know.

**Reply by shibe2** [reply]
Try batch size 512. When you get it working, experiment with different batch sizes.

**Reply by ActuallyTaylor** [reply]
When I set a batch size to 512, I end up receiving this error. Should I be creating multiple batches using `llama_batch_init`?
```
llama.cpp/llama.cpp:5745: n_tokens <= n_batch
```

**Reply by shibe2** [reply]
Maybe call `llama_decode(llama_batch_get_one())` multiple times with no more tokens than the batch size.

**Reply by ActuallyTaylor** [reply]
When using llama_batch_get_one, I am able to decode all of my tokens, but when I attempt to sample I am just being given all whitespace and a token ID of 13. Could you point me in the direction of the right sample strategy? Maybe just the example you use to process large prompts?

**Edit**
If I use the following code:
```
while currentToken <= computedLength {
    let logits = llama_get_logits(context)
    
    let candidatesPointer = UnsafeMutablePointer<llama_token_data>.allocate(capacity: Int(numberOfVocabTokens))
    
    for id in (0..<numberOfVocabTokens) {
        candidatesPointer[Int(id)] = llama_token_data(id: id, logit: logits![Int(id)], p: 0)
    }
    
    var candidates_p: llama_token_data_array = .init(data: candidatesPointer, size: Int(numberOfVocabTokens), sorted: false)
    
    let newTokenID = llama_sample_token(context, &candidates_p)

    guard newTokenID != llama_token_eos(context) && currentToken != computedLength else {
        break
    }
                        
    let piece = try tokenToPiece(token: newTokenID)

    output += piece
    currentToken += 1
}
```

I get basically garbage output that looks something like this:
```
 I But No But They My I Yet I Inc
 Still But Since The So ( This And Well
 But
 When

 I

 They Is They In Is
```

I think my strategy of loading all of the batches by looping over prompt tokens in stride and then using `llama_batch_get_one` to create a batch and then decode with `llama_decode` is leading to some level of corruption that significantly confuses the model.

**Reply by shibe2** [reply]
I'm not familiar with that. Compare your code with examples.

If it works with short prompts (less than 2000 tokens), but outputs garbage for long prompts, one possible problem is that particular model is not capable of handling that amount of tokens with parameters that you use. Make sure that the same exact prompt and the same model file work with examples (main or server).

**Comment by ActuallyTaylor** [comment]
I will try that!

Will setting a batch size to 512 still work through all of the tokens? I am a little confused at how that function works under the hood!

**Reply by shibe2** [reply]
Yes, that is limited by context size. Of course, you need a model that supports large contexts, or at least scaling options.

**Reply by ActuallyTaylor** [reply]
Awesome that is great to know!

**Comment by ActuallyTaylor** [comment]
I partially solved this issue by reimplementing the main.cpp file as a swift implementation. This seems to have solved taking prompts up to 4096 tokens. Anything past this context limit results in the LLM failing and returning garbage data. But for the purpose of this library it has been solved.

**Reply by shibe2** [reply]
You are probably using a model that supports 4096 context and without scaling options.

**Reply by ActuallyTaylor** [reply]
I was indeed only using a model with a context limit of 4096. However, I have changed to a model with 32K recently, hoping for an increase in ability.
